# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/BrowserHost.cpp
# Commit: 405100db6c45
# Full Hash: 405100db6c45f0463286067df5b491f74790d9aa
# Author: Kashav Madan <kmadan@mozilla.com>
# Date: 2019-07-11 09:51:12
# Regressor Bug: 1543251
# File Overlap Count: 1
# Description:
#   Bug 1543251 - Move hasBeforeUnload from PBrowser to PWindowGlobal, r=nika
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D37003
# ==============================================================================

diff -r 0c408343080d -r 405100db6c45 dom/ipc/BrowserHost.cpp
--- a/dom/ipc/BrowserHost.cpp	Thu Jul 04 07:41:57 2019 +0000
+++ b/dom/ipc/BrowserHost.cpp	Wed Jul 10 21:13:44 2019 +0000
@@ -250,15 +250,22 @@
 /* readonly attribute boolean hasBeforeUnload; */
 NS_IMETHODIMP
 BrowserHost::GetHasBeforeUnload(bool* aHasBeforeUnload) {
-  if (!mRoot) {
+  if (!mRoot || !GetBrowsingContext()) {
     *aHasBeforeUnload = false;
     return NS_OK;
   }
+
   bool result = false;
 
-  VisitAll([&result](BrowserParent* aBrowserParent) {
-    result |= aBrowserParent->GetHasBeforeUnload();
-  });
+  GetBrowsingContext()->PreOrderWalk(
+      [&result](BrowsingContext* aBrowsingContext) {
+        WindowGlobalParent* windowGlobal =
+            aBrowsingContext->Canonical()->GetCurrentWindowGlobal();
+
+        if (windowGlobal) {
+          result |= windowGlobal->HasBeforeUnload();
+        }
+      });
 
   *aHasBeforeUnload = result;
   return NS_OK;